src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs

```diff
@@ -584,5 +601,118 @@ protected HqlTreeNode VisitNewArrayExpression(NewArrayExpression expression)
+			if (type == null || toType == null || Equals(type, toType))
+			{
+				existType = false;
+				return false;
+			}
+
+			var sqlTypes = type.SqlTypes(_parameters.SessionFactory);
+			var toSqlTypes = toType.SqlTypes(_parameters.SessionFactory);
+			if (sqlTypes.Length != 1 || toSqlTypes.Length != 1)
+			{
+				existType = false;
+				return false; // Casting a multi-column type is not possible
+			}
+
+			existType = true;
+			if (sqlTypes[0].DbType == toSqlTypes[0].DbType)
+			{
+				return false;
+			}
+
+			if (type.ReturnedClass.IsEnum && sqlTypes[0].DbType == DbType.String)
+			{
+				existType = false;
+				return false; // Never cast an enum that is mapped as string, the type will provide a string for the parameter value
+			}
+
+			// Some dialects can map several sql types into one, cast only if the dialect types are different
+			if (!_parameters.SessionFactory.Dialect.TryGetCastTypeName(sqlTypes[0], out var castTypeName) ||
+			    !_parameters.SessionFactory.Dialect.TryGetCastTypeName(toSqlTypes[0], out var toCastTypeName))
+			{
+				return false; // The dialect does not support such cast
+			}
+
+			return castTypeName != toCastTypeName;
+		}
+
+		private bool IsCastRequired(System.Type type, string sqlFunctionName, out bool existType)
+		{
+			if (type == typeof(object))
+			{
+				existType = false;
+				return false;
+			}
+
+			var toType = TypeFactory.GetDefaultTypeFor(type);
+			if (toType == null)
+			{
+				existType = false;
+				return true; // Fallback to the old behavior
+			}
+
+			existType = true;
+			var sqlFunction = _parameters.SessionFactory.SQLFunctionRegistry.FindSQLFunction(sqlFunctionName);
+			if (sqlFunction == null)
+			{
+				return true; // Fallback to the old behavior
+			}
+
+			var fnReturnType = sqlFunction is ISQLAggregateFunction aggregateFunction
+				? aggregateFunction.GetActualReturnType(toType, _parameters.SessionFactory)
+				: sqlFunction.ReturnType(toType, _parameters.SessionFactory);
+			return fnReturnType == null || IsCastRequired(fnReturnType, toType, out existType);
+		}
+
+		private IType GetType(Expression expression)
+		{
+			if (!(expression is MemberExpression memberExpression))
+			{
+				return expression.Type != typeof(object)
+					? TypeFactory.GetDefaultTypeFor(expression.Type)
+					: null;
+			}
+
+			// Try to get the mapped type for the member as it may be a non default one
+			var entityName = TryGetEntityName(memberExpression);
+			if (entityName == null)
+			{
+				return TypeFactory.GetDefaultTypeFor(expression.Type); // Not mapped
+			}
+
+			var persister = _parameters.SessionFactory.GetEntityPersister(entityName);
+			var index = persister.EntityMetamodel.GetPropertyIndexOrNull(memberExpression.Member.Name);
+			return !index.HasValue
+				? TypeFactory.GetDefaultTypeFor(expression.Type) // Not mapped
+				: persister.EntityMetamodel.PropertyTypes[index.Value];
+		}
+
+		private string TryGetEntityName(MemberExpression memberExpression)
```

Comment:
This is not entirely correct.

Reply:
You are right, I forgot about custom entity names. This method was "enhanced" and moved to `ExpressionsHelper` class as it is also used by #1996, #2036, #2079, but was not updated here. Nonetheless, both methods do not work when having custom entity names. I realized that `IEntityNameProvider` exists and is implemented by the linq query which I will use to calculate the correct entity name instead of using `TryGetGuessEntityName` method. I will update the mentioned PRs when the method is fixed.

Reply:
I took me more than expected as there are a lot mapping features that the method needs to be aware of in order to travese them, but as far as I know the method is now as accurate as it could be. Unfortunately it still needs to guess the entity name when casting is involved (like `((PizzaOrder)o.Order).Name`), but there is nothing we can do about it. I 've added a test for entity names in #2054. 
As this PR has an extension for `ISQLFunction` similar to what am I planning to do in #2061, I will update it after the other one is finished.